### DISCUSSION TITLE
auto run main.py

### DISCUSSION THREAD

**Discussion by antonvh** [pybricks/support#982] [General]
Is there a way to run a script as soon as the PrimeHub/InventorHub boots?

On regular micropython you can call your script main.py or boot.py to achieve this. How to do it with Pybricks?

**Comment by dlech** [comment]
This currently isn't supported (you have to press the button first), but we have discussed adding support for a `boot.py`. 

What is the use case? It seems like this could make a robot unexpectedly run away as soon as it is powered on.

**Reply by rbeckett75** [reply]
I'd like this feature too - I'm in the process of using a spare Move Hub to create a user-operated Great Ball Contraption
I've managed to write some code that used one of the built in motors as a 'timer dial' - the user will turn it, to select how long they want it to run for, then the motor on the other port runs for a certain period of time

Having the program start automatically would just make it easier for the display to start (this is hopefully going in a local Lego shop)
the Shopkeeper can just push a button, to start the program... and visitors can turn the dial to see the GBC run for a few minutes

So no.. I'm not worried about a robot running away unexpectedly :)

**Reply by Stuckalf** [reply]
Me, I'd like this feature, too.
I'm using (and possibly doing a bit of re-work on) Lok23's MotorControl.py for a DCC-like train control of speed and brake.
It would be very handy to not always press the hub button twice. Instead it would be nice to have the script running as soon as the hub has booted.

**Comment by antonvh** [comment]
I want to create an interactive installation with the hub hidden inside the LEGO construction. MINDSTORMS laypeople will operate it, so I'd like it to power on and run as soon as the USB gets power. I suppose Pybricks powers off on SystemExit, after some time with no battery connected.

I'll find more use cases with Bluetooth, but that needs some thinking.

**Reply by laurensvalk** [reply]
For switching off, there is also `hub.system.shutdown()`, and you can monitor the charging state as well:

![image](https://user-images.githubusercontent.com/12326241/224263396-fcd6b31c-c1e2-4d93-8b5a-61bc6d58eec9.png)

![image](https://user-images.githubusercontent.com/12326241/224263190-b6bd7fb8-8762-4c62-8426-948d347fe8e7.png)

**Comment by laurensvalk** [comment]
We have a storage area for [generic user data and settings](https://docs.pybricks.com/en/latest/hubs/primehub.html#pybricks.hubs.PrimeHub.system.storage) on all hubs.

In a similar fashion, we could use a small section of that space for hub configuration. Are there other useful settings we could add if we did this?
- auto_start (True/False), for @antonvh's use case.
- ... ?

**Reply by antonvh** [reply]
- Volume for hubs with a speaker
- programmatical access to hub name

**Reply by antonvh** [reply]
Also
- auto-off time after x seconds of inactivity

**Reply by laurensvalk** [reply]
Is 3 minutes too long/short, or should it be configurable?

**Reply by antonvh** [reply]
3 minutes is great for 99% of use. But for demos and interactives it might be too short. Not super important, but if you're going to do this, I was just adding ideas to the wishlist. 😊

**Reply by EclecticInventor** [reply]
Any update on having a run on boot/power?

**Reply by laurensvalk** [reply]
It isn't planned for the near term.

**Reply by TrainsWithLights** [reply]
Has there been any update on this, given it is now the 'long term' and 2025?
